File: phone/src/protocol/messages/envelope.dart
class Endpoint {
  const Endpoint({required this.deviceId, required this.module});

  final String deviceId;
  final String module;

  Map<String, dynamic> toJson() {
    return {'device_id': deviceId, 'module': module};
  }

  factory Endpoint.fromJson(Map<String, dynamic> json) {
    return Endpoint(
      deviceId: json['device_id'] as String,
      module: json['module'] as String,
    );
  }
}

class Envelope {
  const Envelope({
    required this.messageId,
    required this.traceId,
    required this.messageType,
    required this.messageName,
    required this.protocolVersion,
    required this.source,
    required this.target,
    required this.timestamp,
    required this.payload,
    this.correlationId,
    this.taskId,
    this.sessionId,
    this.priority = 'normal',
    this.requiresAck = false,
    this.metadata = const {},
  });

  final String messageId;
  final String traceId;
  final String messageType;
  final String messageName;
  final String protocolVersion;
  final Endpoint source;
  final Endpoint target;
  final String timestamp;
  final Map<String, dynamic> payload;
  final String? correlationId;
  final String? taskId;
  final String? sessionId;
  final String priority;
  final bool requiresAck;
  final Map<String, dynamic> metadata;

  Map<String, dynamic> toJson() {
    final data = <String, dynamic>{
      'message_id': messageId,
      'trace_id': traceId,
      'message_type': messageType,
      'message_name': messageName,
      'protocol_version': protocolVersion,
      'source': source.toJson(),
      'target': target.toJson(),
      'timestamp': timestamp,
      'payload': payload,
      'priority': priority,
      'requires_ack': requiresAck,
      'metadata': metadata,
    };
    if (correlationId != null) data['correlation_id'] = correlationId;
    if (taskId != null) data['task_id'] = taskId;
    if (sessionId != null) data['session_id'] = sessionId;
    return data;
  }

  factory Envelope.fromJson(Map<String, dynamic> json) {
    return Envelope(
      messageId: json['message_id'] as String,
      traceId: json['trace_id'] as String,
      messageType: json['message_type'] as String,
      messageName: json['message_name'] as String,
      protocolVersion: json['protocol_version'] as String,
      source: Endpoint.fromJson(json['source'] as Map<String, dynamic>),
      target: Endpoint.fromJson(json['target'] as Map<String, dynamic>),
      timestamp: json['timestamp'] as String,
      payload: Map<String, dynamic>.from(json['payload'] as Map? ?? const {}),
      correlationId: json['correlation_id'] as String?,
      taskId: json['task_id'] as String?,
      sessionId: json['session_id'] as String?,
      priority: (json['priority'] as String?) ?? 'normal',
      requiresAck: (json['requires_ack'] as bool?) ?? false,
      metadata: Map<String, dynamic>.from(json['metadata'] as Map? ?? const {}),
    );
  }
}

File: server/src/protocol/messages/envelope.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from protocol.enums import MessageType, Priority


@dataclass(slots=True)
class Endpoint:
    device_id: str
    module: str

    def to_dict(self) -> dict[str, str]:
        return {"device_id": self.device_id, "module": self.module}

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> "Endpoint":
        return cls(device_id=raw["device_id"], module=raw["module"])


@dataclass(slots=True)
class Envelope:
    message_id: str
    trace_id: str
    message_type: MessageType
    message_name: str
    protocol_version: str
    source: Endpoint
    target: Endpoint
    timestamp: str
    payload: dict[str, Any]
    correlation_id: str | None = None
    task_id: str | None = None
    session_id: str | None = None
    priority: Priority = Priority.NORMAL
    requires_ack: bool = False
    metadata: dict[str, Any] = field(default_factory=dict)

    def validate(self) -> None:
        if not self.message_id:
            raise ValueError("message_id is required")
        if not self.trace_id:
            raise ValueError("trace_id is required")
        if "." not in self.message_name:
            raise ValueError("message_name must follow <domain>.<action>")
        if not self.source.device_id or not self.target.device_id:
            raise ValueError("source/target device_id is required")

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {
            "message_id": self.message_id,
            "trace_id": self.trace_id,
            "message_type": self.message_type.value,
            "message_name": self.message_name,
            "protocol_version": self.protocol_version,
            "source": self.source.to_dict(),
            "target": self.target.to_dict(),
            "timestamp": self.timestamp,
            "payload": self.payload,
            "priority": self.priority.value,
            "requires_ack": self.requires_ack,
            "metadata": self.metadata,
        }
        if self.correlation_id:
            data["correlation_id"] = self.correlation_id
        if self.task_id:
            data["task_id"] = self.task_id
        if self.session_id:
            data["session_id"] = self.session_id
        return data

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> "Envelope":
        envelope = cls(
            message_id=raw["message_id"],
            trace_id=raw["trace_id"],
            message_type=MessageType(raw["message_type"]),
            message_name=raw["message_name"],
            protocol_version=raw["protocol_version"],
            source=Endpoint.from_dict(raw["source"]),
            target=Endpoint.from_dict(raw["target"]),
            timestamp=raw["timestamp"],
            payload=dict(raw.get("payload") or {}),
            correlation_id=raw.get("correlation_id"),
            task_id=raw.get("task_id"),
            session_id=raw.get("session_id"),
            priority=Priority(raw.get("priority", Priority.NORMAL.value)),
            requires_ack=bool(raw.get("requires_ack", False)),
            metadata=dict(raw.get("metadata") or {}),
        )
        envelope.validate()
        return envelope



def now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()
